File: Pandacap.Data/TheirContent/DeviantArt/InboxDeviation.fs
namespace Pandacap.Data

open System
open System.ComponentModel.DataAnnotations.Schema
open Pandacap.Types

/// A post from a user who this instance's owner follows on DeviantArt.
[<AbstractClass>]
type InboxDeviation() =
    member val Id = Guid.Empty with get, set
    member val CreatedBy = Guid.Empty with get, set
    member val Username = "" with get, set
    member val Usericon = nullString with get, set
    member val Timestamp = DateTimeOffset.MinValue with get, set
    member val MatureContent = false with get, set
    member val Title = nullString with get, set
    member val LinkUrl = nullString with get, set
    member val DismissedAt = nullDateTimeOffset with get, set

    [<NotMapped>]
    abstract member ThumbnailUrls: string seq

    interface IPost with
        member this.DisplayTitle = this.Title |> orString $"{this.Id}"
        member this.Id = $"{this.Id}"
        member this.LinkUrl = this.LinkUrl
        member this.ProfileUrl = $"https://www.deviantart.com/{Uri.EscapeDataString(this.Username)}"
        member _.Badges = [{ PostPlatform.GetBadge DeviantArt with Text = "www.deviantart.com" }]
        member this.Timestamp = this.Timestamp
        member this.ThumbnailUrls = this.ThumbnailUrls
        member this.Usericon = this.Usericon
        member this.Username = this.Username

File: Pandacap.Database/InboxDeviation.cs
using Pandacap.UI.Badges;
using Pandacap.UI.Elements;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text;

namespace Pandacap.Database
{
    /// <summary>
    /// A post from a user who this instance's owner follows on DeviantArt.
    /// </summary>
    public abstract class InboxDeviation : IInboxPost
    {
        public Guid Id { get; set; }

        public Guid CreatedBy { get; set; }

        public string Username { get; set; } = "";

        public string? Usericon { get; set; }

        public DateTimeOffset Timestamp { get; set; }

        public bool MatureContent { get; set; }

        public string? Title { get; set; }

        public string LinkUrl { get; set; } = "";

        public DateTimeOffset? DismissedAt { get; set; }

        [NotMapped]
        public abstract IEnumerable<IPostThumbnail> Thumbnails { get; }

        bool IInboxPost.IsPodcast => false;

        bool IInboxPost.IsShare => false;

        Badge IPost.Badge => Badges.DeviantArt.WithHostFromUriString(LinkUrl);

        string IPost.DisplayTitle => Title ?? $"{Id}";

        string IPost.Id => $"{Id}";

        string IPost.InternalUrl => LinkUrl;

        string IPost.ExternalUrl => LinkUrl;

        DateTimeOffset IPost.PostedAt => Timestamp;

        string? IPost.ProfileUrl => $"https://www.deviantart.com/{Uri.EscapeDataString(Username)}";

        IEnumerable<IPostThumbnail> IPost.Thumbnails => Thumbnails;
    }
}
